docs/security/index.md

```diff
@@ -0,0 +1,400 @@
+`SecurityLabels` instance or throws a `MalformedLabelsException`.
+
+The `SecurityLabelsParser` to use is obtained via the `SecurityPlugin.labelsParser()` method.  Parsers **SHOULD**
+generally be thread-safe singletons as label byte sequences should be entirely independent of each other and ideally
+require minimal local state to parse.  Implementations *MAY* wish to consider [Caching Parsing
+Results](#caching-for-performance) as in practical deployments we typically see many common labels reused widely across
+the data and having the parser return the same output `SecurityLabels<?>` instance when receiving the same input
+`byte[]` *MAY* allow other parts of the implementation e.g. the [`Authorizer`](#authorizer) to be more efficiently
+implemented.
+
+### `SecurityLabelsValidator`
+
+The `SecurityLabelsValidator` allows an application to take the label byte sequences it has encountered and validate
+them, this is typically intended for applications that are going to store the labels for data their ingesting and won't
+evaluate those labels until later.  By validating the labels applications can actively reject data that is mislabelled
+up front, rather than only discovering the mislabelling later when attempting to grant access to it.
+
+It has a single `validate(byte[])` method that returns a `boolean` indicating whether a given label byte sequence is
+valid for this plugin.
+
+The `SecurityLabelsValidator` to use is obtained via the `SecurityPlugin.labelsValidator()` method.
+
+### `SecurityLabelsApplicator`
+
+The `SecurityLabelsApplicator` allows an application to determine the labels that apply to a specific triple so they can
+store the appropriate label to enable making an access decision at a later time.  This is intended primarily for usage
+when processing a single RDF event from the `knowledge` topic, thus applications obtain instances of these by calling
+the `SecurityPlugin.prepareLabelsApplicator(byte[], Graph)` method providing the labels byte sequence from the
+`Security-Label` header and the fine-grained labels graph (if any).  Therefore plugins **MUST** provide a fresh instance
+of this interface every time their `prepareLabelsApplicator()` method is invoked as the labels that apply **MAY** change
+on every RDF event an application processes.
+
+Once an application has an instance they call `labelForTriple(Triple)` whenever they need to lookup the label for a
+particular triple, this returns a `SecurityLabels<?>` from which the application can then store the corresponding byte
+sequence by accessing its `encoded()` method.
+
+### `Authorizer`
+
+The `Authorizer` allows an application to make access decisions based on the labels for data.  This is intended to be
+scoped to the lifetime of a single user request so an instance is created by calling the
+`SecurityPlugin.prepareAuthorizer(Entitlements<?>)` method passing in the users entitlements (see [Example
+Usage](#example-usage)) for a fuller outline usage.
+
```

Comment:
- [ ] Incomplete description here
- [ ] Make reference to caching for performance again